### PR TITLE
Add example about std.parallelism

### PR DESCRIPTION
**Pull request by wilzbach**
image

### CODE REVIEW

File: index.dd [hunk 1]
```diff
@@ -97,6 +97,24 @@ void main()
...
+$(EXTRA_EXAMPLE
+----
+// Find the logarithm of every number
+// from 1 to 1_000_000 in parallel
+void main()
+{
+    import std.math, std.parallelism, std.stdio;
+
+    auto logs = new double[1_000_000];
+
```
Comment (thewilsonator) [inline index.dd]: Might want to initialise it to iota. Also since the point is to parallelise it maybe add sone timing code to show that it runs faster?

**Comment by CyberShadow** [comment]
Eh, this is a bit boring. More interesting would be to calculate and print an actual result, such as I don't know, like sum of all prime numbers up to a point. Needs some ideas. taskPool.reduce and parallel sorting are also more interesting.

**Comment by quickfur** [comment]
Maybe a benchmark to showcase the difference between parallel and non-parallel foreach.

I'm not too concerned about the example being "interesting" in the sense of doing something complicated / clever; the point of a frontpage example is to show off the language, not the cleverness of the example writer.  I think the key with `std.parallelism` is how easy it is to turn your single-threaded loop into a parallel loop (provided, of course, that loop iterations don't depend on each other).

**Comment by quickfur** [comment]
Here's a rough example that shows the difference:
````
void main()
{
    import std.datetime, std.math, std.parallelism, std.stdio;

    auto logs = new double[1_000_000];
    auto numIterations = 1;

    // Linear init
    auto bm1 = benchmark!({
        foreach (i, ref elem; logs)
        {
            elem = log(1.0 + i);
        }
    })(numIterations);
    writefln("Linear init: %s msecs", bm1[0].msecs);

    // Parallel init
    auto bm2 = benchmark!({
        foreach (i, ref elem; parallel(logs))
        {
            elem = log(1.0 + i);
        }
    })(numIterations);
    writefln("Parallel init: %s msecs", bm2[0].msecs);
}
````

**Comment by quickfur** [comment]
It can probably use some trimming down to size to fit on the front page :D

**Comment by CyberShadow** [comment]
> I'm not too concerned about the example being "interesting" in the sense of doing something complicated / clever; the point of a frontpage example is to show off the language, not the cleverness of the example writer.

Sure, but calculating a result to memory then throwing it away is not great either.

**Comment by quickfur** [comment]
OK, maybe parallel reduce might be more interesting? Like summing an array of 1,000,000 doubles.

**Comment by wilzbach** [comment]
> It can probably use some trimming down to size to fit on the front page :D

Tried to trim it down a bit. We might want to play with the array sizes a bit.
To give you a rough feeling:

```
500_000 x 100:
Linear init: 1639 msecs
Parallel init: 655 msecs

300_000 x 100:
Linear init: 988 msecs
Parallel init: 341 msecs

100_000 x 100:
Linear init: 313 msecs
Parallel init: 118 msecs
```

So the performance of the DLang-Tour backend, isn't that bad ;-)

[Run it yourself](https://tour.dlang.org/editor?source=void%20main()%0A%7B%0A%20%20%20%20import%20std.datetime%2C%20std.math%2C%20std.parallelism%2C%20std.stdio%3B%0A%0A%20%20%20%20auto%20logs%20%3D%20new%20double%5B500_000%5D%3B%0A%20%20%20%20auto%20bm%20%3D%20benchmark!(%7B%0A%20%20%20%20%20%20%20%20foreach%20(i%2C%20ref%20elem%3B%20logs)%0A%20%20%20%20%20%20%20%20%20%20%20%20elem%20%3D%20log(1.0%20%2B%20i)%3B%0A%20%20%20%20%7D%2C%20%7B%0A%20%20%20%20%20%20%20%20foreach%20(i%2C%20ref%20elem%3B%20logs.parallel)%0A%20%20%20%20%20%20%20%20%20%20%20%20elem%20%3D%20log(1.0%20%2B%20i)%3B%0A%20%20%20%20%7D)(100)%3B%20%2F%2F%20number%20of%20repeated%20iterations%0A%20%20%20%20writefln(%22Linear%20init%3A%20%25s%20msecs%22%2C%20bm%5B0%5D.msecs)%3B%0A%20%20%20%20writefln(%22Parallel%20init%3A%20%25s%20msecs%22%2C%20bm%5B1%5D.msecs)%3B%0A%7D).

> Sure, but calculating a result to memory then throwing it away is not great either.

Hmm this (iteration of a list in parallel) is actually my most common use of `std.parallelism`.
The main point to show here is how amazingly easy it is to do parallelism in D.

**Comment by wilzbach** [comment]
So ... imho just showing off how easy & beautiful  `parallel` algorithms can be is a major selling point (and we now easy `benchmark` to prove the point).

@CyberShadow what do you think?

**Comment by CyberShadow** [comment]
Alright.

Lines wrap at the default size:

**Comment by wilzbach** [comment]
Interesting - you seem to use a different font on your FF:

image

**Comment by CyberShadow** [comment]
Oh yeah, `monospace` can mean anything I guess. Okay.

**Comment by CyberShadow** [comment]
> on your FF

Not browser specific BTW

**Comment by wilzbach** [comment]
> monospace can mean anything I guess. Okay.

Btw:

```
font-family: Consolas, "Bitstream Vera Sans Mono", "Andale Mono", Monaco, "DejaVu Sans Mono", "Lucida Console", monospace;
```

I guess you don't have `Consolas` installed?

**Comment by wilzbach** [comment]
Interesting that the bot doesn't merge this ... will investigate.

**Comment by wilzbach** [comment]
I saw the following in the logs:

```
» 20 Jul 2017 02:50:40.247  163 <190>1 2017-07-20T00:50:39.742281+00:00 app web.1 - - [A3313FA1:00000000 dbg] Connection was closed by remote peer (fd 12).
» 20 Jul 2017 02:50:40.247  196 <190>1 2017-07-20T00:50:39.742322+00:00 app web.1 - - [A3313FA1:9C3A0DA1 dbg] Client didn't send the initial request in a timely manner. Closing connection.
```

...

```
 20 Jul 2017 02:52:42.330  137 <190>1 2017-07-20T00:52:41.815660+00:00 app web.1 - - [A3313FA1:9C3A0DA1 dbg] start task (fd 12).
» 20 Jul 2017 02:52:42.330  124 <190>1 2017-07-20T00:52:41.815686+00:00 app web.1 - - [A3313FA1:00000000 dbg] FD: -1
» 20 Jul 2017 02:52:42.330  164 <190>1 2017-07-20T00:52:41.815735+00:00 app web.1 - - [A3313FA1:00000000 dbg] Socket event on fd 12: 17 (33FCA88 vs 33FCA88)
» 20 Jul 2017 02:52:42.330  163 <190>1 2017-07-20T00:52:41.815770+00:00 app web.1 - - [A3313FA1:00000000 dbg] Connection was closed by remote peer (fd 12).
» 20 Jul 2017 02:52:42.330  196 <190>1 2017-07-20T00:52:41.815808+00:00 app web.1 - - [A3313FA1:9C3A0DA1 dbg] Client didn't send the initial request in a timely manner. Closing connection.
» 20 Jul 2017 02:52:42.330  135 <190>1 2017-07-20T00:52:41.815811+00:00 app web.1 - - [A3313FA1:9C3A0DA1 dbg] task out (fd 12).
» 20 Jul 2017 02:52:42.330  132 <190>1 2017-07-20T00:52:41.815817+00:00 app web.1 - - [A3313FA1:9C3A0DA1 dbg] task finished.
```

Will upgrade the bot to Vibe 0.8.0 later -maybe that helps.

**Comment by CyberShadow** [comment]
> I guess you don't have Consolas installed?

That's a Microsoft font and IIRC not a core web font so that's probably it.

**Comment by wilzbach** [comment]
> Will upgrade the bot to Vibe 0.8.0 later -maybe that helps.

FYI: https://github.com/dlang-bots/dlang-bot/pull/137 (deployed the upgrade of the bot to Vibe 0.8.0)

**Pull request merged**